### ISSUE TITLE
Timestamp is rounded on MySQL after save

### ISSUE THREAD

**Issue by ricardograca** [tgriesser/knex#2524]
# Environment

Knex version: 0.14.4
Database + version: Mysql 5.7.21
OS: Ubuntu 16.04 x86_64

# Bug

When using the `.timestamps()` schema builder method on a table, the generated datetime columns will round dates on MySQL when inserting data.

For example, inserting `2018-03-04T19:24:53.063Z` will result in `2018-03-04T19:24:53.000Z` when retrieving the same row afterwards.

Obviously this presents some problems when one wants to compare dates.

[This gist](https://gist.github.com/ricardograca/bc26ee92baeeee3f792a5d90050608cc) has all the code necessary to reproduce the issue. Just make sure the database config matches what you have in your system.

I can look into it and provide a PR.

**Comment by elhigu** [comment]
Have you checked if dates are rounded when before inserting, by database or when results are read from DB? Does running with DEBUG=knex:* environment variable show any light to that?

**Comment by elhigu** [comment]
Also have you tried inserting that date as string `2018-03-04T19:24:53.063Z` to see if the problem is how mysql driver is interpreting `Date` objects.

**Comment by ricardograca** [comment]
I'll investigate all those questions and report back ASAP.

**Comment by ricardograca** [comment]
Seems to be a problem with the actual MySQL server in versions prior to 5.6.4:

> ...when MySQL stores a value into a column of any temporal data type, it discards any fractional part and does not store it.

from [Fractional Seconds in Time Values](https://dev.mysql.com/doc/refman/5.6/en/fractional-seconds.html).

So, it should be possible to change the `.timestamps()` function to output `DATETIME(3)` or `TIMESTAMP(3)` to achieve parity with the other database engines. However, I'm not sure how this will behave in older versions of MySQL. If it's ignored we're good.

**Comment by ricardograca** [comment]
Will you accept a PR for this?

**Comment by elhigu** [comment]
Sorry for being late with this one.

> So, it should be possible to change the .timestamps() function to output DATETIME(3) or TIMESTAMP(3) to achieve parity with the other database engines. However, I'm not sure how this will behave in older versions of MySQL. If it's ignored we're good.

Did you check if it was ignored in older mysql versions? 

For postgresql we have currently some attribute in configuration that can be used to select which version of postgresql server we are connecting to (there was badly working autodetection too, but I prefer saying flavor directly in configuration).

**Comment by ricardograca** [comment]
I checked with MySQL 5.1.69 and it produces a SQL syntax error due to the `(3)` part. However, MySQL 5.6.4 was released on 2011-12-20, which makes it over 6 years old at this point. Not sure if it's really important to support older versions. If it is, then maybe this can be configurable when initializing knex:

```js
var knex = require('knex')({
  client: 'mysql',
  connection: {
    preciseTimestamps: false,
    host : '127.0.0.1',
    user : 'your_database_user',
    password : 'your_database_password',
    database : 'myapp_test'
  }
});
```
I would still prefer to have precise timestamps by default though.

It looks like the `mysql` npm module still tests against 5.5, so I'm going to assume that's the oldest supported version. `node-mysql2` on the other hand only tests against 5.7.

**Comment by elhigu** [comment]
I think that we should always use use precise timestamps. Yes I think that it should be possible to add used dialect version like in postgres. 

So was this supported in all mysql versions after 5.6.4? If I understood correctly mysql docs, 5.6.10 was the first mysql 5.6 series version that was released to general availability (first stable 5.6 release) so versions 5.6.0 - 5.6.9 not working shouldn't be a problem.

I suppose 5.5. is still used in some older server distributions, which has it installed by default. So I don't think that it is obsolete yet (but I don't know how much it is used either... 😐).

I think that safest path would be to keep current functionality by default and add support for telling dialect version when initializing knex and default that version to 5.5 with warning that in future default will be changed to 5.6 (to provide us some path to change default to more reasonable one in future).

This way change will be backwards compatible and people currently using knex with 5.5 will see warning that they need to give version: '5.5' parameter to their config or some future knex versions will not work anymore.

**Comment by ricardograca** [comment]
Well, according to the documentation the feature was introduced with 5.6.4, so I guess it is present in 5.6.10 as well. As a somewhat related note, MariaDB has had support for microsecond precision since version 5.3, so on that side there shouldn't be any problems. 

I'll start working on a PR for this soon.
